// net/Packet.js
"use strict"

module.exports = Packet;

function Packet(){
	this.msgId = 0;
	this.sessionId = 0;
	this.bodySize = 0;
	this.body = null;
}

var prototype = Packet.prototype;

prototype.create = function(){
	return new Packet();
}

prototype.getHeadSize = function(){
	return 28;
}

prototype.getBodySize = function(){
	return this.bodySize;
}

prototype.readHead = function(buffer){
	this.head = buffer.slice(0, this.getHeadSize());
	this.msgId = buffer.readUInt16LE(2);
	this.bodySize = buffer.readUInt32LE(4);
	this.sessionId = buffer.readUInt32LE(16);
}

prototype.readBody = function(buffer){
	//this.body = buffer.toString("utf8", 0, this.bodySize);
	this.body = buffer.slice(0, this.bodySize);
}

prototype.toBytes = function(){
	var bodySize = this.body ? this.body.length : 0;

	var head = new Buffer(this.getHeadSize());
	head.writeUInt16LE(0, 0);
	head.writeUInt16LE(this.msgId, 2);
	head.writeUInt32LE(bodySize, 4);
	head.writeUInt32LE(this.sessionId, 16);

	if(null == this.body){
		return head;
	}
	return Buffer.concat([head, this.body]);
}
